File: fx/src/main/java/gargoyle/fx/FXDialogs.java
package gargoyle.fx;

import gargoyle.fx.icons.Icon;
import gargoyle.fx.log.FXLog;
import javafx.beans.InvalidationListener;
import javafx.scene.Node;
import javafx.scene.Scene;
import javafx.scene.control.Alert;
import javafx.scene.control.Button;
import javafx.scene.control.ButtonBar;
import javafx.scene.control.ButtonType;
import javafx.scene.control.Dialog;
import javafx.scene.control.DialogPane;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.Label;
import javafx.scene.control.TextArea;
import javafx.scene.control.TextInputDialog;
import javafx.scene.image.ImageView;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.util.Callback;

import java.net.URL;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Collectors;

final class FXDialogs {
    private static final String LC_APPLY = "apply";
    private static final String LC_CANCEL = "cancel";
    private static final String LC_CONFIRMATION = "confirmation";
    private static final String LC_ERROR = "error";
    private static final String LC_INFORMATION = "information";
    private static final String LC_LESS = "less";
    private static final String LC_MORE = "more";
    private static final String LC_NO = "no";
    private static final String LC_OK = "ok";
    private static final String LC_PROMPT = "prompt";
    private static final String LC_YES = "yes";
    private static final String MSG_ERROR = "Error";
    private static final String MSG_PROMPT = "Please enter";

    private FXDialogs() {
        throw new IllegalStateException(FXDialogs.class.getName());
    }

    public static Optional<ButtonType> alert(final FXContext context, final Stage owner, final String message,
                                             final ButtonType... buttons) {
        return dialog(context, owner, Alert.AlertType.INFORMATION, message,
                () -> getString(context, LC_INFORMATION, LC_INFORMATION), asMap(buttons));
    }

    public static Optional<ButtonType> alert(final FXContext context, final Stage owner, final String message,
                                             final Map<ButtonBar.ButtonData, String> buttons) {
        return dialog(context, owner, Alert.AlertType.INFORMATION, message,
                () -> getString(context, LC_INFORMATION, LC_INFORMATION), buttons);
    }

    public static ButtonType[] asButtonTypeArray(final Map<ButtonBar.ButtonData, String> buttons) {
        return buttons.entrySet().stream()
                .map(entry -> new ButtonType(entry.getValue(), entry.getKey()))
                .toArray(ButtonType[]::new);
    }

    public static List<ButtonType> asButtonTypeCollection(final Map<ButtonBar.ButtonData, String> buttons) {
        return buttons.entrySet().stream()
                .map(entry -> new ButtonType(entry.getValue(), entry.getKey()))
                .collect(Collectors.toList());
    }

    public static Boolean ask(final FXContext context, final Stage owner, final String message) {
        return confirm(context, owner, message, ButtonType.YES, ButtonType.NO, ButtonType.CANCEL)
                .map(buttonType -> buttonType == ButtonType.YES).orElse(null);
    }

    @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
    public static boolean confirm(final FXContext context, final Stage owner, final String message) {
        return confirm(context, owner, message, ButtonType.OK, ButtonType.CANCEL)
                .map(buttonType -> buttonType == ButtonType.OK).orElse(Boolean.FALSE);
    }

    @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
    public static boolean confirm(final FXContext context, final Stage owner, final String message,
                                  final Map<ButtonBar.ButtonData, String> buttons) {
        return confirmExt(context, owner, message, buttons)
                .map(buttonType -> ButtonBar.ButtonData.OK_DONE == buttonType.getButtonData()).orElse(false);
    }

    public static Optional<ButtonType> confirmExt(final FXContext context,
                                                  final Stage owner, final String message,
                                                  final Map<ButtonBar.ButtonData, String> buttons) {
        return dialog(context, owner, Alert.AlertType.CONFIRMATION, message,
                () -> getString(context, LC_CONFIRMATION, LC_CONFIRMATION), buttons);
    }

    public static Hyperlink createHyperlink(final String text, final Consumer<? super Hyperlink> action) {
        final Hyperlink hyperlink = new Hyperlink(text);
        hyperlink.setOnAction(event -> action.accept(hyperlink));
        return hyperlink;
    }

    public static <R> void decorateDialog(final FXContext context, final Dialog<R> dialog,
                                          final Callback<ButtonType, R> resultConverter,
                                          final Map<ButtonBar.ButtonData, String> buttons,
                                          final String title) {
        decorateDialog(context, dialog, resultConverter, buttons, title,
                (isExpanded, detailsButton) -> detailsButton.setText(isExpanded ?
                        getString(context, LC_LESS, LC_LESS) : getString(context, LC_MORE, LC_MORE)));
    }

    public static <R> void decorateDialog(final FXContext context, final Dialog<R> dialog,
                                          final Callback<ButtonType, R> resultConverter,
                                          final Map<ButtonBar.ButtonData, String> buttons,
                                          final String title,
                                          final BiConsumer<? super Boolean, ? super Hyperlink> detailsButtonDecorator) {
        dialog.setTitle(title);
        dialog.setDialogPane(new DecoratedDialogPane((buttonData, button) -> Icon.find(buttonData).findIcon(context)
                .map(URL::toExternalForm).map(ImageView::new)
                .ifPresent(button::setGraphic),
                detailsButtonDecorator));
        if (null == buttons || buttons.isEmpty()) {
            dialog.getDialogPane().getButtonTypes().setAll(ButtonType.OK, ButtonType.CANCEL);
        } else {
            dialog.getDialogPane().getButtonTypes().setAll(asButtonTypeCollection(buttons));
        }
        dialog.setResultConverter(resultConverter);
    }

    public static void decorateDialogAs(final FXContext context, final Stage owner, final Alert dialog,
                                        final Map<ButtonBar.ButtonData, String> buttons) {
        decorateDialogAs(owner, dialog, (buttonData, button) -> {
                    Icon.find(buttonData).findIcon(context)
                            .map(URL::toExternalForm).map(ImageView::new)
                            .ifPresent(button::setGraphic);
                    if (null == buttons || buttons.isEmpty() || !buttons.containsKey(buttonData)) {
                        button.setText(translateButton(context, buttonData, button.getText()));
                    }
                },
                (isExpanded, detailsButton) -> detailsButton.setText(isExpanded ?
                        getString(context, LC_LESS, LC_LESS) : getString(context, LC_MORE, LC_MORE)));
    }

    public static Optional<ButtonType> dialog(final FXContext context,
                                              final Stage owner,
                                              final Alert.AlertType type,
                                              final String message,
                                              final Supplier<String> titleProvider,
                                              final Map<ButtonBar.ButtonData, String> buttons) {
        final Alert dialog = null == buttons || buttons.isEmpty() ?
                new Alert(type, message) :
                new Alert(type, message, asButtonTypeArray(buttons));
        decorateDialogAs(context, owner, dialog, buttons);
        dialog.setTitle(getTitle(owner, titleProvider));
        dialog.setHeaderText(null);
        dialog.setContentText(message);
        setModal(owner, dialog);
        return dialog.showAndWait();
    }

    public static Optional<ButtonType> error(final FXContext context, final Stage owner,
                                             final String message, final Exception ex,
                                             final ButtonType... buttons) {
        return error(context, owner, message, ex, asMap(buttons));
    }

    public static Optional<ButtonType> error(final FXContext context, final Stage owner,
                                             final String message, final Exception ex,
                                             final Map<ButtonBar.ButtonData, String> buttons) {
        if (null != ex) {
            FXLog.info(ex, ex.getLocalizedMessage());
        }
        final Alert dialog = new Alert(Alert.AlertType.ERROR);
        decorateDialogAs(context, owner, dialog, buttons);
        dialog.setTitle(getTitle(context, owner, LC_ERROR, MSG_ERROR));
        dialog.setHeaderText(null);
        dialog.setContentText(message);
        if (null != ex) {
            final Label label = new Label(ex.getClass().getName());
            final String exceptionText = FXUtil.stringStackTrace(ex);
            final TextArea textArea = new TextArea(exceptionText);
            textArea.setEditable(false);
            textArea.setWrapText(true);
            textArea.setMaxWidth(Double.MAX_VALUE);
            textArea.setMaxHeight(Double.MAX_VALUE);
            GridPane.setVgrow(textArea, Priority.ALWAYS);
            GridPane.setHgrow(textArea, Priority.ALWAYS);
            final GridPane expContent = new GridPane();
            expContent.setMaxWidth(Double.MAX_VALUE);
            expContent.add(label, 0, 0);
            expContent.add(textArea, 0, 1);
            dialog.getDialogPane().setExpandableContent(expContent);
        }
        if (null != buttons && !buttons.isEmpty()) {
            dialog.getButtonTypes().setAll(asButtonTypeCollection(buttons));
        }
        setModal(owner, dialog);
        return dialog.showAndWait();
    }

    public static Optional<ButtonType> error(final FXContext context, final Stage owner, final String message,
                                             final ButtonType... buttons) {
        return error(context, owner, message, asMap(buttons));
    }

    public static Optional<ButtonType> error(final FXContext context, final Stage owner, final String message,
                                             final Map<ButtonBar.ButtonData, String> buttons) {
        final Alert dialog = new Alert(Alert.AlertType.INFORMATION);
        decorateDialogAs(context, owner, dialog, buttons);
        dialog.setTitle(getTitle(context, owner, LC_ERROR, MSG_ERROR));
        dialog.setHeaderText(null);
        dialog.setContentText(message);
        if (null != buttons && !buttons.isEmpty()) {
            dialog.getButtonTypes().setAll(asButtonTypeCollection(buttons));
        }
        setModal(owner, dialog);
        return dialog.showAndWait();
    }

    public static String prompt(final FXContext context, final Stage owner, final String message) {
        return prompt(context, owner, message, "");
    }

    private static Map<ButtonBar.ButtonData, String> asMap(final ButtonType[] buttons) {
        return Arrays.stream(buttons).collect(Collectors.toMap(ButtonType::getButtonData, ButtonType::getText));
    }

    private static Optional<ButtonType> confirm(final FXContext context,
                                                final Stage owner, final String message, final ButtonType... buttons) {
        return dialog(context, owner, Alert.AlertType.CONFIRMATION, message,
                () -> getString(context, LC_CONFIRMATION, LC_CONFIRMATION), asMap(buttons));
    }

    private static void decorateDialogAs(final Stage primaryStage, final Dialog<?> dialog,
                                         final BiConsumer<? super ButtonBar.ButtonData, ? super Button> buttonDecorator,
                                         final BiConsumer<? super Boolean, ? super Hyperlink> detailsButtonDecorator) {
        decorateDialogButtons(dialog, buttonDecorator, detailsButtonDecorator);
        Optional.ofNullable(dialog).map(Dialog::getDialogPane).map(Node::getScene).map(Scene::getWindow)
                .filter(Stage.class::isInstance).map(Stage.class::cast)
                .filter(stage -> Objects.nonNull(primaryStage))
                .ifPresent(stage -> doDecorateStageAs(primaryStage, stage));
    }

    private static void decorateDialogButtons(final Dialog<?> dialog,
                                              final BiConsumer<? super ButtonBar.ButtonData, ? super Button> buttonDecorator,
                                              final BiConsumer<? super Boolean, ? super Hyperlink> detailsButtonDecorator) {
        Optional.ofNullable(dialog).ifPresent(window -> {
            final DialogPane oldDialogPane = window.getDialogPane();
            window.setDialogPane(new DecoratedDialogPane(buttonDecorator, detailsButtonDecorator));
            window.getDialogPane().getButtonTypes().setAll(oldDialogPane.getButtonTypes());
        });
    }

    private static void doDecorateStageAs(final Stage primaryStage, final Stage window) {
        window.setTitle(primaryStage.getTitle());
        window.getIcons().setAll(primaryStage.getIcons());
    }

    private static String getString(final FXContext context,
                                    final String key,
                                    final String defaultTitle) {
        return FXUtil.message(FXDialogs.class, context.getLocale(), () -> defaultTitle, key, Map.of());
    }

    private static String getTitle(final Stage owner, final Supplier<String> titleProvider) {
        return Optional.ofNullable(owner).map(Stage::getTitle).orElseGet(titleProvider);
    }

    private static String getTitle(final FXContext context,
                                   final Stage owner, final String key, final String defaultTitle) {
        return getTitle(owner, () -> getString(context, key, defaultTitle));
    }

    @SuppressWarnings("SameParameterValue")
    private static String prompt(final FXContext context,
                                 final Stage owner, final String message, final String defaultValue) {
        final TextInputDialog dialog = new TextInputDialog(defaultValue);
        decorateDialogAs(owner, dialog, (buttonData, button) -> {
                    Icon.find(buttonData).findIcon(context).map(URL::toExternalForm).map(ImageView::new)
                            .ifPresent(button::setGraphic);
                    button.setText(translateButton(context, buttonData, button.getText()));
                },
                (isExpanded, detailsButton) -> detailsButton.setText(isExpanded ?
                        getString(context, LC_LESS, LC_LESS) : getString(context, LC_MORE, LC_MORE)));
        dialog.setTitle(getTitle(context, owner, LC_PROMPT, MSG_PROMPT));
        dialog.setHeaderText(null);
        dialog.setContentText(message);
        setModal(owner, dialog);
        return dialog.showAndWait().orElse(null);
    }

    private static void setModal(final Stage owner, final Dialog<?> dialog) {
        Optional.ofNullable(owner).ifPresent(stage -> {
            dialog.initOwner(stage);
            dialog.initModality(Modality.WINDOW_MODAL);
        });
    }

    private static String translateButton(final FXContext context,
                                          final ButtonBar.ButtonData buttonData, final String defaultText) {
        return switch (buttonData) {
            case OK_DONE -> getString(context, LC_OK, LC_OK);
            case CANCEL_CLOSE -> getString(context, LC_CANCEL, LC_CANCEL);
            case YES -> getString(context, LC_YES, LC_YES);
            case NO -> getString(context, LC_NO, LC_NO);
            case APPLY -> getString(context, LC_APPLY, LC_APPLY);
            default -> defaultText;
        };
    }

    private static final class DecoratedDialogPane extends DialogPane {
        private static final String CLASS_DETAILS_BUTTON = "details-button";
        private static final String CLASS_DETAILS_BUTTON_LESS = "less";
        private static final String CLASS_DETAILS_BUTTON_MORE = "more";
        private static final String LESS_TEXT = "Dialog.detail.button.less";
        private static final String MORE_TEXT = "Dialog.detail.button.more";
        private final BiConsumer<? super ButtonBar.ButtonData, ? super Button> buttonDecorator;
        private final BiConsumer<? super Boolean, ? super Hyperlink> detailsButtonDecorator;

        private DecoratedDialogPane(final BiConsumer<? super ButtonBar.ButtonData, ? super Button> buttonDecorator,
                                    final BiConsumer<? super Boolean, ? super Hyperlink> detailsButtonDecorator) {
            this.buttonDecorator = buttonDecorator;
            this.detailsButtonDecorator = detailsButtonDecorator;
        }

        @Override
        protected Node createButton(final ButtonType buttonType) {
            final Node node = super.createButton(buttonType);
            if (node instanceof Button) {
                buttonDecorator.accept(buttonType.getButtonData(), (Button) node);
            }
            return node;
        }

        @Override
        protected Node createDetailsButton() {
            final Hyperlink detailsButton = new Hyperlink();
            final InvalidationListener expandedListener = observable -> {
                final boolean isExpanded = isExpanded();
                detailsButton.getStyleClass().setAll(CLASS_DETAILS_BUTTON,
                        isExpanded ? CLASS_DETAILS_BUTTON_LESS : CLASS_DETAILS_BUTTON_MORE);
                detailsButton.setText(isExpanded ? LESS_TEXT : MORE_TEXT);
                detailsButtonDecorator.accept(isExpanded, detailsButton);
            };
            expandedListener.invalidated(null);
            expandedProperty().addListener(expandedListener);
            detailsButton.setOnAction(ae -> setExpanded(!isExpanded()));
            return detailsButton;
        }
    }
}

File: fx/src/main/kotlin/gargoyle/fx/FxDialogs.kt
package gargoyle.fx

import gargoyle.fx.icons.Icon
import gargoyle.fx.log.FxLog
import javafx.beans.InvalidationListener
import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.control.Alert
import javafx.scene.control.Alert.AlertType
import javafx.scene.control.Button
import javafx.scene.control.ButtonBar.ButtonData
import javafx.scene.control.ButtonType
import javafx.scene.control.Dialog
import javafx.scene.control.DialogPane
import javafx.scene.control.Hyperlink
import javafx.scene.control.Label
import javafx.scene.control.TextArea
import javafx.scene.control.TextInputDialog
import javafx.scene.image.ImageView
import javafx.scene.layout.GridPane
import javafx.scene.layout.Priority
import javafx.stage.Modality
import javafx.stage.Stage
import javafx.util.Callback

@Suppress("SameParameterValue", "MemberVisibilityCanBePrivate")
internal object FxDialogs {

    private class DecoratedDialogPane(
        private val buttonDecorator: (ButtonData, Button) -> Unit,
        private val detailsButtonDecorator: (Boolean, Hyperlink) -> Unit
    ) : DialogPane() {
        override fun createButton(buttonType: ButtonType): Node =
            super.createButton(buttonType)
                .also { if (it is Button) buttonDecorator(buttonType.buttonData, it) }

        override fun createDetailsButton(): Node {
            val detailsButton = Hyperlink()
            val expandedListener = InvalidationListener {
                detailsButton.styleClass.setAll(
                    CLASS_DETAILS_BUTTON,
                    if (this.isExpanded) CLASS_DETAILS_BUTTON_LESS else CLASS_DETAILS_BUTTON_MORE
                )
                detailsButton.text = if (this.isExpanded) LESS_TEXT else MORE_TEXT
                detailsButtonDecorator(this.isExpanded, detailsButton)
            }
            expandedListener.invalidated(null)
            expandedProperty().addListener(expandedListener)
            detailsButton.onAction = EventHandler { isExpanded = !isExpanded }
            return detailsButton
        }

        companion object {
            private const val CLASS_DETAILS_BUTTON = "details-button"
            private const val CLASS_DETAILS_BUTTON_LESS = "less"
            private const val CLASS_DETAILS_BUTTON_MORE = "more"
            private const val LESS_TEXT = "Dialog.detail.button.less"
            private const val MORE_TEXT = "Dialog.detail.button.more"
        }
    }

    private const val LC_APPLY = "apply"
    private const val LC_CANCEL = "cancel"
    private const val LC_CONFIRMATION = "confirmation"
    private const val LC_ERROR = "error"
    private const val LC_INFORMATION = "information"
    private const val LC_LESS = "less"
    private const val LC_MORE = "more"
    private const val LC_NO = "no"
    private const val LC_OK = "ok"
    private const val LC_PROMPT = "prompt"
    private const val LC_YES = "yes"
    private const val MSG_ERROR = "Error"
    private const val MSG_PROMPT = "Please enter"

    fun alert(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Array<ButtonType> = arrayOf()
    ): ButtonType? =
        dialog(
            message,
            AlertType.INFORMATION,
            context,
            owner,
            { getString(LC_INFORMATION, LC_INFORMATION, context) },
            buttons.associate { it.buttonData to it.text }
        )

    fun alert(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Map<ButtonData, String> = mapOf()
    ): ButtonType? =
        dialog(
            message,
            AlertType.INFORMATION,
            context,
            owner,
            { getString(LC_INFORMATION, LC_INFORMATION, context) },
            buttons
        )

    fun ask(message: String, context: FxContext = FxContext.current, owner: Stage = context.stage): Boolean? =
        confirm(
            message,
            context,
            owner,
            arrayOf(ButtonType.YES, ButtonType.NO, ButtonType.CANCEL)
        )?.let { it == ButtonType.YES }

    fun confirm(message: String, context: FxContext = FxContext.current, owner: Stage = context.stage): Boolean =
        confirm(
            message,
            context,
            owner,
            arrayOf(ButtonType.OK, ButtonType.CANCEL)
        )?.let { it == ButtonType.OK } ?: false

    fun confirm(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Map<ButtonData, String> = mapOf()
    ): Boolean =
        confirmExt(
            message,
            context,
            owner,
            buttons
        )?.let { ButtonData.OK_DONE == it.buttonData } ?: false

    fun confirmExt(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Map<ButtonData, String> = mapOf()
    ): ButtonType? =
        dialog(
            message,
            AlertType.CONFIRMATION,
            context,
            owner,
            { getString(LC_CONFIRMATION, LC_CONFIRMATION, context) },
            buttons
        )

    fun createHyperlink(text: String, action: (Hyperlink) -> Unit): Hyperlink =
        Hyperlink(text).also { hyperlink -> hyperlink.onAction = EventHandler { action(hyperlink) } }

    fun <R> decorateDialog(
        context: FxContext,
        dialog: Dialog<R>,
        resultConverter: (ButtonType) -> R?,
        buttons: Map<ButtonData, String>,
        title: String
    ) =
        decorateDialog(dialog, resultConverter, title, context, buttons) { expanded: Boolean, button: Hyperlink ->
            button.text = if (expanded) getString(LC_LESS, LC_LESS, context) else getString(LC_MORE, LC_MORE, context)
        }

    fun <R : Any?> decorateDialog(
        dialog: Dialog<R>,
        resultConverter: (ButtonType) -> R?,
        title: String = "",
        context: FxContext = FxContext.current,
        buttons: Map<ButtonData, String> = mapOf(),
        detailsButtonDecorator: (Boolean, Hyperlink) -> Unit
    ) {
        with(dialog) {
            this.title = title
            dialogPane = DecoratedDialogPane(
                { buttonData: ButtonData, button: Button ->
                    Icon.find(buttonData)
                        .findIcon(context)?.toExternalForm()
                        ?.let { ImageView(it) }
                        ?.also { button.graphic = it }
                }, detailsButtonDecorator
            )
            dialogPane.buttonTypes.setAll(
                if (buttons.isEmpty()) listOf(ButtonType.OK, ButtonType.CANCEL)
                else buttons.map { (key, value) -> ButtonType(value, key) }
            )
            this.resultConverter = Callback(resultConverter)
        }
    }

    fun decorateDialogAs(
        dialog: Alert,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Map<ButtonData, String> = mapOf()
    ) =
        decorateDialogAs(dialog, owner, { buttonData: ButtonData, button: Button ->
            Icon.find(buttonData).findIcon(context)?.toExternalForm()
                ?.let { ImageView(it) }
                ?.let { button.graphic = it }
            if (buttons.isEmpty() || !buttons.containsKey(buttonData)) {
                button.text = translateButton(buttonData, button.text, context)
            }
        }) { expanded: Boolean, button: Hyperlink ->
            button.text = if (expanded) getString(LC_LESS, LC_LESS, context) else getString(LC_MORE, LC_MORE, context)
        }

    fun dialog(
        message: String,
        type: AlertType = AlertType.INFORMATION,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        titleProvider: () -> String,
        buttons: Map<ButtonData, String> = mapOf()
    ): ButtonType? =
        with(
            if (buttons.isEmpty()) Alert(type, message) else Alert(
                type,
                message,
                *buttons.map { (key, value) -> ButtonType(value, key) }.toTypedArray()
            )
        ) {
            decorateDialogAs(this, context, owner, buttons)
            title = getTitle(owner, titleProvider)
            headerText = null
            contentText = message
            setModal(owner, this)
            this.showAndWait()
        }.orElse(null)

    fun error(
        message: String,
        ex: Exception? = null,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Array<ButtonType> = arrayOf()
    ): ButtonType? =
        error(message, ex, context, owner, buttons.associate { it.buttonData to it.text })

    fun error(
        message: String,
        ex: Exception? = null,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Map<ButtonData, String> = mapOf()
    ): ButtonType? {
        ex?.let { FxLog.info(ex.localizedMessage, ex) }
        with(Alert(AlertType.ERROR)) {
            decorateDialogAs(this, context, owner, buttons)
            title = getTitle(LC_ERROR, MSG_ERROR, context, owner)
            headerText = null
            contentText = message
            ex?.let {
                with(TextArea(ex.stackTraceToString())) {
                    isEditable = false
                    isWrapText = true
                    maxWidth = Double.MAX_VALUE
                    maxHeight = Double.MAX_VALUE
                    GridPane.setVgrow(this, Priority.ALWAYS)
                    GridPane.setHgrow(this, Priority.ALWAYS)
                    GridPane().also {
                        with(it) {
                            maxWidth = Double.MAX_VALUE
                            add(Label(ex.javaClass.name), 0, 0)
                            add(this, 0, 1)
                        }
                    }.let { dialogPane.expandableContent = it }
                }
            }
            if (buttons.isNotEmpty()) buttonTypes.setAll(buttons.map { (key, value) -> ButtonType(value, key) })
            setModal(owner, this)
            return showAndWait().orElse(null)
        }
    }

    fun error(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Array<ButtonType> = arrayOf()
    ): ButtonType? =
        error(message, context, owner, buttons.associate { it.buttonData to it.text })

    fun error(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Map<ButtonData, String> = mapOf()
    ): ButtonType? =
        with(Alert(AlertType.INFORMATION)) {
            decorateDialogAs(this, context, owner, buttons)
            title = getTitle(LC_ERROR, MSG_ERROR, context, owner)
            headerText = null
            contentText = message
            if (buttons.isNotEmpty()) buttonTypes.setAll(buttons.map { (key, value) -> ButtonType(value, key) })
            setModal(owner, this)
            showAndWait()
        }.orElse(null)

    fun prompt(message: String, context: FxContext = FxContext.current, owner: Stage = context.stage): String =
        prompt(message, "", context, owner)

    private fun confirm(
        message: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage,
        buttons: Array<ButtonType> = arrayOf()
    ): ButtonType? =
        dialog(
            message,
            AlertType.CONFIRMATION,
            context,
            owner,
            { getString(LC_CONFIRMATION, LC_CONFIRMATION, context) },
            buttons.associate { it.buttonData to it.text }
        )

    private fun decorateDialogAs(
        dialog: Dialog<*>,
        primaryStage: Stage = FxContext.current.stage,
        buttonDecorator: (ButtonData, Button) -> Unit,
        detailsButtonDecorator: (Boolean, Hyperlink) -> Unit
    ) {
        decorateDialogButtons(dialog, buttonDecorator, detailsButtonDecorator)
        (dialog.dialogPane.scene?.window)?.let { doDecorateStageAs(primaryStage, it as Stage) }
    }

    private fun decorateDialogButtons(
        dialog: Dialog<*>,
        buttonDecorator: (ButtonData, Button) -> Unit,
        detailsButtonDecorator: (Boolean, Hyperlink) -> Unit
    ) {
        dialog.dialogPane.let {
            with(dialog) {
                dialogPane = DecoratedDialogPane(buttonDecorator, detailsButtonDecorator)
                dialogPane.buttonTypes.setAll(it.buttonTypes)
            }
        }
    }

    private fun doDecorateStageAs(primaryStage: Stage = FxContext.current.stage, window: Stage) {
        with(window) {
            title = primaryStage.title
            icons.setAll(primaryStage.icons)
        }
    }

    private fun getString(key: String, defaultTitle: String = key, context: FxContext = FxContext.current): String =
        FxUtil.message(FxDialogs::class, context.locale, { defaultTitle }, key, mapOf())

    private fun getTitle(owner: Stage, titleProvider: () -> String): String =
        owner.title ?: titleProvider()

    private fun getTitle(
        key: String,
        defaultTitle: String = key,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage
    ): String =
        getTitle(owner) { getString(key, defaultTitle, context) }

    private fun prompt(
        message: String,
        defaultValue: String,
        context: FxContext = FxContext.current,
        owner: Stage = context.stage
    ): String =
        with(TextInputDialog(defaultValue)) {
            decorateDialogAs(this, owner, { buttonData: ButtonData, button: Button ->
                Icon.find(buttonData)
                    .findIcon(context)?.toExternalForm()
                    ?.let { ImageView(it) }
                    ?.let { button.graphic = it }
                button.text = translateButton(buttonData, button.text, context)
            }) { expanded: Boolean, button: Hyperlink ->
                button.text =
                    if (expanded) getString(LC_LESS, LC_LESS, context) else getString(LC_MORE, LC_MORE, context)
            }
            title = getTitle(LC_PROMPT, MSG_PROMPT, context, owner)
            headerText = null
            contentText = message
            setModal(owner, this)
            showAndWait()
        }.orElse(null)

    private fun setModal(owner: Stage, dialog: Dialog<*>) =
        with(dialog) {
            initOwner(owner)
            initModality(Modality.WINDOW_MODAL)
        }

    private fun translateButton(
        buttonData: ButtonData,
        defaultText: String = "",
        context: FxContext = FxContext.current
    ): String =
        when (buttonData) {
            ButtonData.OK_DONE -> getString(LC_OK, LC_OK, context)
            ButtonData.CANCEL_CLOSE -> getString(LC_CANCEL, LC_CANCEL, context)
            ButtonData.YES -> getString(LC_YES, LC_YES, context)
            ButtonData.NO -> getString(LC_NO, LC_NO, context)
            ButtonData.APPLY -> getString(LC_APPLY, LC_APPLY, context)
            else -> defaultText
        }

}
